src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs

```diff
@@ -0,0 +1,184 @@
+                    if (!lookupResult.IsSingleViable)
+                    {
+                        diagnostics.Add(
+                            ErrorCode.ERR_WithParameterWithoutMatchingMember,
+                            syntax.Receiver.Location,
+                            receiverType,
+                            p.Name);
+                        member = null;
+                    }
+                    else
+                    {
+                        member = lookupResult.SingleSymbolOrDefault;
+                        Debug.Assert(member.Kind == SymbolKind.Field || member.Kind == SymbolKind.Property);
+
+                        if (!member.GetTypeOrReturnType().Equals(
+                            p.TypeWithAnnotations,
+                            TypeCompareKind.ConsiderEverything))
+                        {
+                            diagnostics.Add(
+                                ErrorCode.ERR_WithParameterTypeDoesntMatchMemberType,
+                                syntax.Receiver.Location,
+                                p.Name,
+                                p.Type,
+                                member.GetTypeOrReturnType());
+                        }
+                    }
+
+                    lookupResult.Clear();
+                    matchingMembers.Add(member);
+                }
+                withMembers = matchingMembers.ToImmutableAndFree();
+            }
+
+            var args = ArrayBuilder<(Symbol?, BoundExpression)>.GetInstance();
+            // Bind with expression arguments
+            foreach (var initializer in syntax.Initializers)
+            {
+                var propName = initializer.NameEquals?.Name.Identifier.Text;
+                Symbol? member = null;
+                if (propName is object)
+                {
+                    this.LookupMembersInType(
+                        lookupResult,
+                        receiverType,
+                        propName,
+                        arity: 0,
+                        basesBeingResolved: null,
+                        options: LookupOptions.Default, // properties are not invocable - their accessors are
+                        originalBinder: this,
+                        diagnose: false,
+                        useSiteDiagnostics: ref useSiteDiagnostics);
+                    if (lookupResult.IsSingleViable)
+                    {
+                        member = lookupResult.SingleSymbolOrDefault;
+                        Debug.Assert(member.Kind == SymbolKind.Field || member.Kind == SymbolKind.Property);
+                    }
+                    else
+                    {
+                        diagnostics.Add(initializer.NameEquals!.Name, useSiteDiagnostics);
+                    }
+                }
+
+                var expr = BindValue(initializer.Expression, diagnostics, BindValueKind.RValue);
+                lookupResult.Clear();
+                args.Add((member, expr));
```

Comment:
>member [](start = 26, length = 6)

Requires a check that the member is an instance member.  If not corresponding to a parameter of the With method, there should be a requirement that a property has an accessible setter.